text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,1036 @@
+}
+```
+
+Allowing `arc` to be safely moved out of `KeepAlive<T>` would create the false impression that it is
+safe to use `arc` — it is not. By requiring `unsafe` to read `arc`, Rust's safety tooling ensures a
+narrow safety boundary: the user is forced to justify their actions when accessing `arc` (which
+documents its safety conditions as they relate to `KeepAlive`), rather than in downstream
+interactions with `UnsafeCell<T>` (whose methods necessarily provide only general guidance).
+Consequently, we require that moving unsafe fields out of their enclosing type requires `unsafe`. 
+
+### Field Copying is Safe
+
+We propose that all uses of unsafe fields require `unsafe`, including copying. Alternatively, we
+might consider making field copies safe. However, a field may carry an invariant that could be
+violated as consequence a copy. For example, consider a field of type `&'static RefCell<T>` that
+imposes an invariant on the value of `T`. In this alternative proposal, such a field could be safely
+copiable out of its enclosing type, then safely mutated via the API of `RefCell`. Consequently, we
+require that copying unsafe fields out of their enclosing type requires `unsafe`.
+
+### Copy Is Safe To Implement
+
+We propose that `Copy` is conditionally unsafe to implement; i.e., that the `unsafe` modifier is
+required to implement `Copy` for types that have unsafe fields. Alternatively, we can imagine
+permitting retaining Rust's present behavior that `Copy` is unconditionally safe to implement for
+all types; e.g.:
+
+```rust
+struct UnalignedMut<'a, T> {
+    /// # Safety
+    ///
+    /// `ptr` is an exclusive reference to a valid-but-unaligned instance of `T`.
+    unsafe ptr: *mut T,
+    _lifetime: PhantomData<&'a T>,
+}
+
+impl<'a, T> Copy for UnalignedMut<'a, T> {}
+
+impl<'a, T> Clone for UnalignedMut<'a, T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+```
+
+However, the `ptr` field introduces a declaration-site safety obligation that is not discharged
+with `unsafe` at any use site; this violates [**Tenet: Unsafe Usage is Always
+Unsafe**](#tenet-unsafe-usage-is-always-unsafe).
+
+### Non-Trivial Destructors are Prohibited
+
+If a programmer applies the `unsafe` modifier to a field with a non-trivial destructor and relaxes
+its invariant beyond that which is required by the field's destructor, Rust cannot prevent the
+unsound use of that field in safe contexts. This is, seemingly, a soft violation of [**Tenet: Unsafe
+Usage is Always Unsafe**](#tenet-unsafe-usage-is-always-unsafe). We resolve this by documenting that
+such fields are a serious violation of good safety hygiene, and accept the risk that this
+documentation is ignored. This risk is minimized by prevalence: we feel that relaxing a field's
+invariant beyond that of its destructor is a rare subset of the cases in which a field carries a
+relaxed variant, which itself a rare subset of the cases in which a field carries a safety
+invariant.
```

Comment:
@RalfJung, I want to double check with you that nothing stated here violates the model, as you see it, for `unsafe` in Rust and doesn't, as far as you can tell, create any new problems for us.